src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+            "bot_frame": ["  "] * (self._qwires + self._cwires),
+        }
+
+    def _adjust_layer(self, wire_list: List[int], xskip: int):
+        """
+        Adjust the layers by filling the empty spaces with respective characters
+        """
+
+        for wire in wire_list:
+            self._render_strs["top_frame"][wire] += " " * (
+                xskip - len(self._render_strs["top_frame"][wire])
+            )
+            self._render_strs["bot_frame"][wire] += " " * (
+                xskip - len(self._render_strs["bot_frame"][wire])
+            )
+            if wire < self._qwires:
+                self._render_strs["mid_frame"][wire] += "─" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+            else:
+                self._render_strs["mid_frame"][wire] += "═" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+
+    def _add_wire_labels(self):
+        """
+        Add wire labels to the circuit
+        """
+
+        default_labels = [f"q{i}" for i in range(self._qwires)] + [
+            f"c{i}" for i in range(self._cwires)
+        ]
+        for i, label in enumerate(default_labels):
+            self._render_strs["mid_frame"][i] = (
+                f" {label} :" + self._render_strs["mid_frame"][i]
+            )
+            self._render_strs["top_frame"][i] = " " * (len(f" {label} :") + 2)
+            self._render_strs["bot_frame"][i] = " " * (len(f" {label} :") + 2)
+
+        self._layer_list = {
+            i: [len(self._render_strs["mid_frame"][i])]
+            for i in range(self._qwires + self._cwires)
+        }
+
+    def _extend_line(self):
+        """
+        Extend all the wires to the same length
+        """
+
+        wire_len = [sum(wire) for wire in self._layer_list.values()]
+        max_len = max(wire_len) + self.style.end_wire_ext
+
+        for i, length in enumerate(wire_len):
+            if length < max_len:
+                diff = max_len - length
+                if i < self._qwires:
+                    self._render_strs["mid_frame"][i] += "─" * diff
+                else:
+                    self._render_strs["mid_frame"][i] += "═" * diff
+
+    def _draw_singleq_gate(self, gate_name: str):
+        """
+        Draw a single qubit gate
+
+        Parameters
+        ----------
+        gate : Gate
+            The gate to be drawn.
+
+        Returns
+        -------
+        tuple
+            The parts of the gate to be drawn. The parts are the top_frame, mid_frame, and bot_frame.
+
+        int
+            The width of the gate.
+        """
+
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_name))
+        pad = " " * ceil(self.style.gate_pad)
+
+        top_frame = f" ┌{lid_seg}┐ "
+        mid_frame = f"─┤{pad}{gate_name}{pad}├─"
+        bot_frame = f" └{lid_seg}┘ "
+
+        return (top_frame, mid_frame, bot_frame), len(top_frame)
```

Comment:
Should we add an assert checking that the length of all three should be the same? What do you think? This could even replace the last two tests.

Reply:
It can replace one test, `test_parts_len`, and the other test checks if all `_render_str` or wire lengths are consistent. I'm okay with either option; the choice might come down to which provides better coverage, while both options ensure testing. Let me know what you think.

Reply:
It won't affect the coverage because the general test will also cover these lines of code I think.

The point is test_parts_len is to find some specific error and make debugging easier. If it is equivalent to adding assert here. I think it is better to have assert here because it helps detect problems early, where the cause is clear, rather than later when some other operation fails.